File: 0208-implement-trie-prefix-tree/0208-implement-trie-prefix-tree.cpp
struct Node {
    Node *links[26];
    int countEndsWith = 0;
    int countPrefix = 0;
    
    bool containsKey(char ch) {
        return (links[ch - 'a'] != nullptr);
    }

    Node* get(char ch) {
        return links[ch - 'a'];
    }

    void put(char ch, Node *node) {
        links[ch - 'a'] = node;
    }

    void increaseEnd() {
        countEndsWith++;
    }

    void increasePrefix() {
        countPrefix++;
    }

    void reduceEnd() {
        countEndsWith--;
    }

    void reducePrefix() {
        countPrefix--;
    }

    int getEndsWith() {
        return countEndsWith;
    }

    int getPrefix() {
        return countPrefix;
    }
};

class Trie {
private:
    Node *root;
public:
    Trie() {
        ios_base::sync_with_stdio(false); 
        cin.tie(NULL);
        root = new Node();
    }
    
    void insert(string word) {
        Node *node = root;
        for (int i = 0; i < word.size(); i++) {
            if (!node->containsKey(word[i])) {
                node->put(word[i], new Node());
            }
            node = node->get(word[i]);
            node->increasePrefix();
        }
        node->increaseEnd();
    }
    
    bool search(string word) {
        Node *node = root;
        for (int i = 0; i < word.size(); i++) {
            if (node->containsKey(word[i])) {
                node = node->get(word[i]);
            } else {
                return 0;
            }
        }
        return node->getEndsWith();
    }
    
    bool startsWith(string word) {
        Node *node = root;
        for (int i = 0; i < word.size(); i++) {
            if (node->containsKey(word[i])){
                node = node->get(word[i]);
            } else {
                return 0;
            }
        }
        return node->getPrefix();
    }
};

/**
 * Your Trie object will be instantiated and called as such:
 * Trie* obj = new Trie();
 * obj->insert(word);
 * bool param_2 = obj->search(word);
 * bool param_3 = obj->startsWith(prefix);
 */

File: 0208-implement-trie-prefix-tree/0208-implement-trie-prefix-tree.java
class Trie {
    Node root;

    public Trie() {
        root = new Node();
    }
    
    public void insert(String word) {
        Node curr = root;
        for (char ch : word.toCharArray()) {
            if (curr.child[ch - 'a'] == null) {
                curr.child[ch - 'a'] = new Node();
            }

            curr = curr.child[ch - 'a'];
        }

        curr.endsCount += 1;
    }
    
    public boolean search(String word) {
        Node curr = root;
        for (char ch : word.toCharArray()) {
            if (curr.child[ch - 'a'] == null) {
                return false;
            }

            curr = curr.child[ch - 'a'];
        }

        if (curr.endsCount >= 1) {
            return true;
        }

        return false;
    }
    
    public boolean startsWith(String prefix) {
        Node curr = root;
        for (char ch : prefix.toCharArray()) {
            if (curr.child[ch - 'a'] == null) {
                return false;
            }

            curr = curr.child[ch - 'a'];
        }

        return true;
    }

    class Node {
        Node[] child;
        int endsCount;

        public Node() {
            child = new Node[26];
            endsCount = 0;
        }
    }
}

/**
 * Your Trie object will be instantiated and called as such:
 * Trie obj = new Trie();
 * obj.insert(word);
 * boolean param_2 = obj.search(word);
 * boolean param_3 = obj.startsWith(prefix);
 */